### PR TITLE
Gene Finder Code Submission

### PR DESCRIPTION
**Pull request by hmowen**

### CODE REVIEW

File: gene_finder.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 """
 YOUR HEADER COMMENT HERE
```
Comment (LucyWilcox) [inline gene_finder.py]: You should include a header comment here and mention what your code does.

File: gene_finder.py [hunk 2]
```diff
@@ -21,18 +20,30 @@ def shuffle_string(s):
...
-
+	""" Returns the complementary nucleotide
+
+	    nucleotide: a nucleotide (A, C, G, or T) represented as a string
+	    returns: the complementary nucleotide
+	>>> get_complement('A')
+	'T'
+	>>> get_complement('C')
+	'G'
+	>>> get_complement('R')
```
Comment (LucyWilcox) [inline gene_finder.py]: Good additional doc tests, but maybe comment on what/why you added them.

File: gene_finder.py [hunk 3]
```diff
@@ -44,44 +55,72 @@ def get_reverse_complement(dna):
...
-    pass
-
+    rdna = dna[::-1]																		# reverses input
 
-def rest_of_ORF(dna):
-    """ Takes a DNA sequence that is assumed to begin with a start
-        codon and returns the sequence up to but not including the
-        first in frame stop codon.  If there is no in frame stop codon,
-        returns the whole string.
+    rev_dna = ""																			# initializes empty string
```
Comment (LucyWilcox) [inline gene_finder.py]: rdna and rev_dna could have better names. Also if you are normally using underscores you should stick with that.

File: gene_finder.py [hunk 4]
```diff
@@ -44,44 +55,72 @@ def get_reverse_complement(dna):
...
+	dna: a DNA sequence
+	returns: the open reading frame represented as a string
+	>>> rest_of_ORF("ATGTGAA")
+	'ATG'
+	>>> rest_of_ORF("ATGAGATAGG")
+	'ATGAGA'
+	>>> rest_of_ORF("ATGTAGTAGTAG")					# Edge case
+	'ATG'
+	"""
+	# TODO: implement this
```
Comment (LucyWilcox) [inline gene_finder.py]: Remove TODO flag.

File: gene_finder.py [hunk 5]
```diff
@@ -44,44 +55,72 @@ def get_reverse_complement(dna):
...
+	index = 0
+	while index < len(dna):
+		codon = dna[index:index + 3]														# groups 3 nucleotides to make codon, index by 3
+		if codon in ["TAA", "TAG", "TGA"]:													# checks if stop codon is in string
+			return dna[:index]																# if True, returns string up to stop codon
+		index += 3																			# indexes by 3
+	return dna
+
+#print rest_of_ORF("ATGAGATAGG")
```
Comment (LucyWilcox) [inline gene_finder.py]: Delete commented out line, but it's not a big deal right now.

File: gene_finder.py [hunk 6]
```diff
@@ -44,44 +55,72 @@ def get_reverse_complement(dna):
...
+	returns the whole string.
+
+	dna: a DNA sequence
+	returns: the open reading frame represented as a string
+	>>> rest_of_ORF("ATGTGAA")
+	'ATG'
+	>>> rest_of_ORF("ATGAGATAGG")
+	'ATGAGA'
+	>>> rest_of_ORF("ATGTAGTAGTAG")					# Edge case
+	'ATG'
```
Comment (LucyWilcox) [inline gene_finder.py]: What about a test case where there is no end? You do catch this in your code though!

File: gene_finder.py [hunk 7]
```diff
@@ -119,8 +169,14 @@ def longest_ORF(dna):
     >>> longest_ORF("ATGCGAATGTAGCATCAAA")
     'ATGCTACATTCGCAT'
     """
-    # TODO: implement this
-    pass
+    longest_ORF = []																		# initialize empty list
+    longest_ORF = longest_ORF + find_all_ORFs_both_strands(dna)								# adds all ORFs both strands to list
```
Comment (LucyWilcox) [inline gene_finder.py]: You could just initialize here to longer_ORF = find_all_ORFs_both_strands(dna). You don't need to initialize the empty list first.

File: gene_finder.py [hunk 8]
```diff
@@ -119,8 +169,14 @@ def longest_ORF(dna):
...
+
+    for n in range(len(longest_ORF)):														# searches through the list
```
Comment (LucyWilcox) [inline gene_finder.py]: Here you're looping to check if something is equal to max(longest_ORF, key=len) then returning the element if it is. Instead of doing this, you should just return max(longest_ORF, key=len)

File: gene_finder.py [hunk 9]
```diff
@@ -130,9 +186,13 @@ def longest_ORF_noncoding(dna, num_trials):
         dna: a DNA sequence
         num_trials: the number of random shuffles
         returns: the maximum length longest ORF """
-    # TODO: implement this
-    pass
+    longest_random_ORF = []																	# initialize empty list
+    for i in range(num_trials):																# repeats for num_trials
+    	randomized_dna = shuffle_string(dna)												# finds strings of random dna
+    	longest_random_ORF.append(longest_ORF(randomized_dna))								# adds random dna to list
+	return max(longest_random_ORF, key=len)													# returns longest dna string
```
Comment (LucyWilcox) [inline gene_finder.py]: This return needs to be outside of the for loop, right now you are returning the orf from the first trial, not the longest.
Also this function is meant to return a length, not an orf.

File: gene_finder.py [hunk 10]
```diff
@@ -148,9 +208,13 @@ def coding_strand_to_AA(dna):
         >>> coding_strand_to_AA("ATGCCCGCTTT")
         'MPA'
     """
-    # TODO: implement this
-    pass
+    s = ""																					# intitialize empty list
```
Comment (LucyWilcox) [inline gene_finder.py]: You should chose a better name for this.

File: gene_finder.py [hunk 11]
```diff
@@ -148,9 +208,13 @@ def coding_strand_to_AA(dna):
         >>> coding_strand_to_AA("ATGCCCGCTTT")
         'MPA'
     """
-    # TODO: implement this
-    pass
+    s = ""																					# intitialize empty list																		
+    for i in range(0, len(dna)-2, 3):														# for range of length of dna, indexes w/ step 3 (to isolate codons)
```
Comment (LucyWilcox) [inline gene_finder.py]: Maybe comment on why you have the number 2 here. You explain three in the comment, but not two. Also if you use three a lot you can use a global variable for it like: DNA_ORF_CONVERSION_LEN = 3

File: gene_finder.py [hunk 12]
```diff
@@ -158,8 +222,24 @@ def gene_finder(dna):
...
-    pass
+    viable_strings = []																		# intitialize empty list (for strings)
+    viable_amino_acids = []																	# intitialize empty list (for amino acids)
+    threshold = longest_ORF_noncoding(dna, 1500)											# sets threshold to longest random dna string
+    real_dna = list(find_all_ORFs_both_strands(dna))										# sets real_dna equal to all the ORFs, both strands
+    for i in range(len(real_dna)):															# searches through all the elements in list real_dna
+    	if len(real_dna[i]) > len(threshold):												# compares real string to random string
+    		viable_strings.append(real_dna[i])												# if real string is longer, adds it to list
+    for i in range(len(viable_strings)):													# searches through all elements in viable_strings
+    	a = coding_strand_to_AA(viable_strings[i])											# translates each string to amino acid sequence
```
Comment (LucyWilcox) [inline gene_finder.py]: Use a better name than a.

File: gene_finder.py [hunk 13]
```diff
@@ -158,8 +222,24 @@ def gene_finder(dna):
...
+    	viable_amino_acids.append(a)														# adds amino acids to list
+    return viable_amino_acids																# returns list
+
+
+
+
+dna = load_seq("./data/X73525.fa")
+
+print gene_finder(dna)
```
Comment (LucyWilcox) [inline gene_finder.py]: This and the dna = load_seq(...
should be in you if __name __ = "__main__":
